sled-agent/src/zone_bundle.rs

```diff
@@ -1545,61 +1545,54 @@ async fn compute_bundle_utilization(
 
 // Return the number of bytes occupied by the provided directory.
 //
-// This returns an error if:
-//
-// - The "du" command fails
-// - Parsing stdout fails
-// - Parsing the actual size as a u64 fails
+// This returns an error if reading any file or directory within the provided
+// directory fails.
 async fn disk_usage(path: &Utf8PathBuf) -> Result<u64, BundleError> {
-    // Each OS implements slightly different `du` options.
-    //
-    // Linux and illumos support the "apparent" size in bytes, though using
-    // different options. macOS doesn't support bytes at all, and has a minimum
-    // block size of 512.
-    //
-    // We'll suffer the lower resolution on macOS, and get higher resolution on
-    // the others.
-    cfg_if::cfg_if! {
-        if #[cfg(target_os = "illumos")] {
-            const BLOCK_SIZE: u64 = 1;
-            const DU_ARG: &str = "-A";
-        } else if #[cfg(target_os = "linux")] {
-            const BLOCK_SIZE: u64 = 1;
-            const DU_ARG: &str = "-b";
-        } else if #[cfg(target_os = "macos")] {
-            const BLOCK_SIZE: u64 = 512;
-            const DU_ARG: &str = "-k";
-        } else {
-            compile_error!("unsupported target OS");
+    let path = path.clone();
+    // We could, alternatively, just implement this using `tokio::fs` to read
```

Comment:
I have not looked at this much yet but I wonder if we should just use the `walkdir` crate that we already have?  I feel like there are so many edge cases in doing traversal that it's worth using a crate (but I don't feel strongly).

Reply:
ah, i didn't realize that was in our dependency tree, I'll take a look at it.

Reply:
b4e873109e033d68d0dbd804cb693ace27a96485 changes this to use `walkdir`. I will add that, on my machine, the `walkdir` implementation takes ~100 ms to calculate the size of `sled-agent/src`, while my previous implementation did it in ~8 ms, and `du(1)` only takes about 6 ms. I'm not sure if we care about performance here that much, though, so that may not matter a whole lot --- making sure we're handling edge cases is definitely important, but we're also not using all that much of `walkdir`'s functionality (e.g. traversing symlinks).

Reply:
Looking at https://github.com/uutils/coreutils/blob/8fe93ded741f8257e9d0a5bce12f904e09227a59/src/uu/du/src/du.rs#L319, at least `du` looks pretty simple. There are definitely issues around traversing symlinks and bind mounts (loops in general) but I guess the only other check one could do is to ensure we don't cross filesystems.

Reply:
Also that code's recursive, huh. I'd definitely write this using a heap-allocated stack personally.

Reply:
> Also that code's recursive, huh. I'd definitely write this using a heap-allocated stack personally.

That was what my original implementation did! :)